pkg/sql/sem/tree/eval.go

```diff
@@ -417,9 +417,48 @@ func initArrayToArrayConcatenation() {
 	}
 }
 
+// initNonArrayToNonArrayConcatenation initializes string + nonarrayelement
+// and nonarrayelement + string concatenation.
+func initNonArrayToNonArrayConcatenation() {
```

Comment:
can you add a test that checks `<anyNonArrayElem> || NULL = anyNonArrayElem.String()` for each non array elem type?

Reply:
i may not be following, but in the operation you posted, the result should be NULL.

i added a test to `pkg/sql/sem/tree/testdata/eval/concat`. See https://github.com/rafiss/cockroach/blob/3705e3c7f37effaa644d45344b5c89a461877401/pkg/sql/sem/tree/testdata/eval/concat#L73 

i can add more tests for the other non array types.

Reply:
ah woops, i meant anyNonArrayElem || string, as a loop over all current types to make sure it doesn't get resolved unexpectedly.

Reply:
it didn't seem very nice to loop over types and generate values for each type, so instead i just added more cases to `concat` eval test. it turned up an issue so i updated eval.go to use the FmtBareStrings flag

Reply:
> it didn't seem very nice to loop over types and generate values for each type

(i'm not opposed to this landing this as is, but curious)
does `RandDatum` not give you something nice?

Reply:
i didn't know about `RandDatum` hehehe. thank you for the pointer!

will use that, and anyway looks like i need to fix tuple-string concat:

```
 encountered internal error:
no volatility for cast tuple{int, int}::tuple
(1) assertion failure
Wraps: (2) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql/opt/memo.BuildSharedProps
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:1512
```